### DISCUSSION TITLE
Equivalent for glob ** for selective sync?

### DISCUSSION THREAD

**Discussion by Interpause** [abraunegg/onedrive#2222] [Q&A] [answered]
Here is an example folder structure:

```
|-foo
| |-a
| | |-bar
| |-b
| | |-bar
| |-c
|   |-d
|     |-bar
|-zaz
  |-bar
  |-e
    |-bar
```

I want to ignore `bar` under `foo` only. Typically, I would write the glob pattern as `!/foo/**/bar`, which would ignore `/foo/a/bar`, `/foo/b/bar` and `/foo/c/d/bar`, while still syncing `bar` inside of `/zaz`.

How would I go about doing this?

**Comment by abraunegg** [comment, marked as answer]
@Interpause 
Please read https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md as this is clearly explained.

**Reply by Interpause** [reply]
```
# Include all Work/Project directories
Work/Project*
```

Does this mean doing `Work/Project` ignores all descendant `Project` folders, not just the immediate child `Project` folder? The example isn't fully clear to me.

**Reply by abraunegg** [reply]
@Interpause 

> Does this mean doing `Work/Project` ignores all descendant `Project` folders, not just the immediate child `Project` folder? The example isn't fully clear to me.

The examples are pretty clear ....

> I want to ignore `bar` under `foo` only. Typically, I would write the glob pattern as `!/foo/**/bar/`, which would ignore `/foo/a/bar`, `/foo/b/bar` and `/foo/c/d/bar`, while still syncing `bar` inside of `/zaz`.

Directives can be path explicit or relative. This is all relative to the 'sync_dir' path.

You have 2 options:

* Use 'skip_dir' using an entry like `/foo/*/bar/` to exclude `/foo/{a,b,c,1,2,3,.....}/bar`. This should however still include `/zaz/bar` as the path being skipped is an explicit match.
*  Use 'sync_list' and be explicit in what you want to be excluded:
```
!/foo/*/bar/
```

Again, use `--display-config` to validate your configuration and use `--dry-run` to validate your configuration is performing what you expect it to do.

**Reply by Interpause** [reply]
I see, so multi-level isn't possible then. Thanks.

**Reply by abraunegg** [reply]
@Interpause 
> I see, so multi-level isn't possible then. Thanks.

Incorrect. Configuring the client to skip directories at one level, whilst allowing the same name at others is *entirely* possible. 

I have given you also the *correct* configuration for your `/foo/bar/` example ...... 

Configure the client, use `--display-config` to validate your configuration and use `--dry-run` to validate your configuration is performing what you expect it to do.

**Reply by Interpause** [reply]
I see it is well possible to do stuff such as `a/*/*/*/c`. To give an example of what I mean by multi-level, lets say I have `.trash` folders nested at different folder levels that I want to ignore only in `/old_notes` but not in `/new_notes`.

**Reply by abraunegg** [reply]
@Interpause 
> I see it is well possible to do stuff such as `a/*/*/*/c`. To give an example of what I mean by multi-level, lets say I have `.trash` folders nested at different folder levels that I want to ignore only in `/old_notes` but not in `/new_notes`.

So you write the appropriate rule to exclude | ignore the required directory in the required ignore locations. It is 100% possible to do this.

Please read:
* https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#how-to-skip-directories-from-syncing
* https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#skip_dir

Again, configure the client, use `--display-config` to validate your configuration and use `--dry-run` to validate your configuration is performing what you expect it to do ....

**Reply by Interpause** [reply]
To make it clear, `.trash` folders are created whenever I delete notes within the folder. The level of nesting is dynamic as I may create new folders for notes within existing folders while organizing them. This would mean modifying the `sync_list` whenever I re-organize my folders. This would also require a resync each time. The reason I wish to do `**` is to avoid writing multiple `*/*/...` rules for every single folder that may exist at some point in the future within the notes folder.

I am also unable to just ignore `.trash` globally. In my example, I want to ignore `.trash` only within the `/old_notes` folder, not the `/new_notes` folder.

**Reply by abraunegg** [reply]
@Interpause 
Have you *actually* tried the solution given ?

Use 'skip_dir' rather than 'sync_list' ... 

Please configure and test your scenarios .........

**Reply by Interpause** [reply]
Ah, I see, sorry for the trouble. So if the * wildcard behaves as according to the Microsoft documentation linked there, doing `notes/*/.trash` should ignore all descendants _only_ if using `skip_dir` instead of `sync_list`. And if I wanted to ignore only the immediate descendant, I would use `notes/*/.trash` again but in `sync_list`. Thats quite inconsistent though.

**Reply by abraunegg** [reply]
> Ah, I see, sorry for the trouble. So if the * wildcard behaves as according to the Microsoft documentation linked there, doing `notes/*/.trash` should ignore all descendants _only_ if using `skip_dir` instead of `sync_list`. And if I wanted to ignore only the immediate descendant, I would use `notes/*/.trash` again but in `sync_list`. Thats quite inconsistent though.

Ummm .. there is no inconsistency here.

If you use 'sync_list' or 'skip_dir' the *exact same* operations would occur.

The only difference in 'sync_list' is you would *exclude* the directories with `!` or just specify all the directories you *want* to sync.

There is *zero* inconsistency.

**Reply by Interpause** [reply]
If that is the case, then there is no way to ignore `a/*/c` without ignoring `a/*/*/c`, `a/*/*/*/c` and so on? Unless explicitly including `a/*/*/c` would work. I think there is a reason other pattern matchers implement `**` else `*` should be able to everything. Would it be possible to implement pattern matching similar to `.gitignore`, which is already fairly rudimentary? Or perhaps, even support for using regex matching?

**Reply by abraunegg** [reply]
@Interpause 
You are *very confused* .....

Please stop trying to paper test the application. What you want to do can be done .. you are just either refusing to test and try .. or have some other motive here. 

Configure and test the application for your desired outcome.
